DQ_SerialManipulator Abstract Class

juanjqo opened the pull request:
Changes before we can update:

### General

- [ ]  Remove legacy pre-release files.

 

- [ ] Make Doxygen documentation automatically build and made available online (C++ Doxygen documentation #1).

### DQ_SerialManipulator

 

- [x] Add get_{lower,upper}_{q,q_dot}_limit

- [x]  Add set_{lower,upper}_{q,q_dot}_limit

 

- [x] Make it abstract.

### DQ_SerialManipulatorDH

 

- [x] Update it to support the changes in DQ_SerialManipulator

- [x]  Unify the {get,set}_{thetas, as, ds, alphas, types}

- [ ]  Make a compatible pose_jacobian_derivative

- [x]  Deprecate the DQ_SerialManipulatorDH(MatrixXd, string) constructor.

### DQ_HolonomicBase

 

- [ ] Fix The pose_jacobian method of DQ_holonomicBase class returns a wrong Jacobian [BUG] python#19

mmmarinho commented:
@juanjqo Please merge this pull request of yours with the current version of the master branch.

(so that it will be automatically checked by GithubActions).

mmmarinho commented:
@juanjqo I’m trying to figure out why this pull request isn’t getting tested by GithubActions.
I modified the workflow a bit, so for now I have to possible explanations:
1. You need to merge again your local branch with the workflow modifications I made.
2. We need to restart this pull request (keep your branch alive and there will be no problems).

Please try 1. and if it doesn’t work we move to 2.

mmmarinho commented:
@juanjqo Thanks!

As expected, the modifications will make some of the examples stop working. Please propose the necessary changes in the examples.

juanjqo commented:
Hi @mmmarinho!

Thank you for your support. I'm going to check the examples.

juanjqo commented:
@mmmarinho, I did the modifications for the Matlab examples in this [pull request](https://github.com/dqrobotics/matlab-examples/pull/1). The new modifications are compatible with [this version](https://github.com/dqrobotics/matlab/pull/69) of the DQ Robotics.

mmmarinho commented:
Converting back to draft until it passes the tests.

mmmarinho commented:
@juanjqo Please roll-back your changes to 1f9f224, and let's get this pull request reviewed.

juanjqo commented:
> @juanjqo Please roll-back your changes to [1f9f224](https://github.com/dqrobotics/matlab/commit/1f9f22426a00cdae0c19930008d082fd55a13a9c), and let's get this pull request reviewed.

@mmmarinho, I reverted the last three commits. Furthermore, I switched this PR to "Ready for review".

mmmarinho commented on robot_modeling/DQ_SerialManipulator.m:
```diff
@@ -13,15 +13,30 @@
 %       set_effector - Set an arbitrary end-effector rigid transformation with respect to the last frame in the kinematic chain.
+%       get_lower_q_limit - Returns the lower bound of the robot configuration.
+%       set_lower_q_limit - Sets the lower bound of the robot configuration.
+%       get_lower_q_dot_limit () - Returns the lower bound of the robot configuration velocity.
```
Some methods have brackets and some don't. Is there a specific reason for this or perhaps just a typo?

mmmarinho commented on robot_modeling/DQ_SerialManipulator.m:
```diff
@@ -13,15 +13,30 @@
+%       get_lower_q_limit - Returns the lower bound of the robot configuration.
+%       set_lower_q_limit - Sets the lower bound of the robot configuration.
+%       get_lower_q_dot_limit () - Returns the lower bound of the robot configuration velocity.
+%       set_lower_q_dot_limit () - Seturns the lower bound of the robot configuration velocity.
```
Brackets

mmmarinho commented on robot_modeling/DQ_SerialManipulator.m:
```diff
@@ -13,15 +13,30 @@
+%       set_lower_q_dot_limit () - Seturns the lower bound of the robot configuration velocity.
+%       get_upper_q_limit - Returns the upper bound of the robot configuration.
+%       set_upper_q_limit - Sets the upper bound of the robot configuration.
+%       get_upper_q_dot_limit () - Returns the upper bound of the robot configuration velocity.
```
Brackets

mmmarinho commented on robot_modeling/DQ_SerialManipulator.m:
```diff
@@ -13,15 +13,30 @@
+%       get_upper_q_limit - Returns the upper bound of the robot configuration.
+%       set_upper_q_limit - Sets the upper bound of the robot configuration.
+%       get_upper_q_dot_limit () - Returns the upper bound of the robot configuration velocity.
+%       set_upper_q_dot_limit () - Seturns the upper bound of the robot configuration velocity.
```
Brackets

mmmarinho commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,73 @@
+
+Type in your terminal:
+
+- `git clone https://github.com/you_user/matlab.git`
```
`your_user
`

juanjqo replied:
@mmmarinho, modified the explanation about how to clone the forked repository. I removed "your_user".

mmmarinho reviewed (changes requested):
@juanjqo 
A few minor suggestions for improvements.

bvadorno commented on robot_modeling/DQ_SerialManipulatorDH.m:
```diff
@@ -202,13 +160,152 @@
-        function w = get_w(obj,ith)       
-            if obj.type(ith) == obj.JOINT_ROTATIONAL
-                w = DQ.k;
+        function ds = get_ds(obj)
```
get_dh_parameters_d

bvadorno commented on robot_modeling/DQ_SerialManipulatorDH.m:
```diff
@@ -202,13 +160,152 @@
+            ds = obj.dh_matrix_(2,:); %obj.d;
+        end
+        
+        function as = get_as(obj)
```
get_dh_parameters_a

Please, do the same for the other parameters.

bvadorno commented on robot_modeling/DQ_SerialManipulatorDH.m:
```diff
@@ -202,13 +160,152 @@
+            alphas =  obj.dh_matrix_(4,:); %obj.alpha;            
+        end
+        
+        function types = get_types(obj)
```
get_joint_types

bvadorno commented on robot_modeling/DQ_SerialManipulatorDH.m:
```diff
@@ -202,13 +160,152 @@
+        
+        function types = get_types(obj)
+            % GET_TYPES() returns the fifth row of the Matrix dh_matrix_, which
+            % correspond to the type of joints of the robot: 
```
which corresponds

juanjqo replied:
According to some requested modifications, now we have:
`% GET_JOINT_TYPES() Returns the joint type, which can be either REVOLUTE or PRISMATIC.`
This is temporal since the name of the joints (I mean, JOINT_REVOLUTE, REVOLUTE_JOINT, REVOLUTE, etc) has not been chosen yet.

bvadorno replied:
Since MATLAB has an [enumeration class](https://uk.mathworks.com/help/matlab/enumeration-classes.html), it suffices to create a class called JointType, and then we can have just the types REVOLUTE and PRISMATIC.

bvadorno commented on robot_modeling/DQ_SerialManipulatorDH.m:
```diff
@@ -202,13 +160,152 @@
+            %obj.convention = convention;
+            %obj.n_links = size(A,2);
+              
+            obj = obj@DQ_SerialManipulator(size(A,2));
```
Why do we need to use an argument in this constructor if the dimension of the configuration space can be inferred from the length of the matrix? Does it have to do with the way we must do in Python and C++?

juanjqo replied:
I am not sure if the argument in the constructor of DQ_SerialManipulator() is strictly required. I implemented it as it is in C++. @mmmarinho do you have any comments?

bvadorno commented on robot_modeling/DQ_SerialManipulatorDH.m:
```diff
@@ -202,13 +160,152 @@
+                error('Input: Invalid DH matrix. It should have 5 rows.')
+            end
+            
+            % Add type
```
remove unnecessary comments

bvadorno commented on robot_modeling/DQ_SerialManipulatorDH.m:
```diff
@@ -202,13 +160,152 @@
+                vec_x_effector_dot = J*q_dot;
+            end
+            
+            obj.check_to_ith_link(n);
```
This check should be within the else statement.

bvadorno commented on robot_modeling/DQ_SerialManipulatorDH.m:
```diff
@@ -224,12 +321,15 @@
             
             if nargin < 3
-                to_ith_link = obj.n_links;
+                to_ith_link = obj.dim_configuration_space_;
```
In order to "simplify" the code, a bad practice was done here, which does not necessarily translate well to different languages: if there are fewer than three parameters, it is assumed that the parameter `to_ith_link` will not be used, hence it is safe to instantiate a local variable with the same name. This practice is confusing, hence it is better to use the previous implementation.

mmmarinho replied:
@bvadorno That was AFAIK introduced by me in 20.04 and slipped through our review. I didn’t know it was considered a bad practice in MATLAB, so I’ll keep that in mind. 

In other languages, it would be handled as follows. Regardless of what I wrote below, we must keep the code consistent so @juanjqo please fix this one and double check if I haven’t used the same logic in other places. 

- It is commonplace in Python to default the argument as None and give it a value inside the method for arguments such as the one in question that cannot be evaluated when the module is loaded. (e.g., some of the examples in Section 2.12.4 https://google.github.io/styleguide/pyguide.html)
- In CPP we do not need to do this because we can overload the method, and that’s how it’s implemented there.

juanjqo replied:
@bvadorno @mmmarinho I did the requested modifications. Now, the method raw_pose_jacobian looks like:

```Matlab
            obj.check_q_vec(q);
            if nargin == 3
                n = to_ith_link;
            else
                n = obj.dim_configuration_space_;
            end

            obj.check_to_ith_link(n);
            x_effector = obj.raw_fkm(q,n);
          
            x = DQ(1);
            J = zeros(8,n);
            
            for i = 1:n
                w = obj.get_w(i);
                z = 0.5*Ad(x,w);
                x = x*obj.dh2dq(q(i),i);
                j = z * x_effector;
                J(:,i) = vec8(j);
            end
```

bvadorno commented on robot_modeling/DQ_SerialManipulatorDH.m:
```diff
@@ -224,12 +321,15 @@
             end
             x_effector = obj.raw_fkm(q,to_ith_link);
             
+            obj.check_to_ith_link(to_ith_link);
```
Place this check before the first use of `to_ith_link` in line 329.

bvadorno commented on robots/ComauSmartSixRobot.m:
```diff
@@ -32,14 +32,16 @@
             comau_DH_d =     [-0.45, 0, 0, -0.64707, 0, -0.095];
             comau_DH_a =     [0, 0.150, 0.590, 0.13, 0, 0];
             comau_DH_alpha = [pi, pi/2, pi, -pi/2, -pi/2, pi/2];
+            comau_DH_type = repmat(DQ_SerialManipulatorDH.JOINT_ROTATIONAL,1,6);
```
Fix according to my suggestions in previous files.

bvadorno commented on robots/FrankaEmikaPandaRobot.m:
```diff
@@ -0,0 +1,61 @@
+% (C) Copyright 2011-2019 DQ Robotics Developers
```
Change the date.

bvadorno commented on robots/FrankaEmikaPandaRobot.m:
```diff
@@ -0,0 +1,61 @@
+%     You should have received a copy of the GNU Lesser General Public License
+%     along with DQ Robotics.  If not, see <http://www.gnu.org/licenses/>.
+%
+% DQ Robotics website: dqrobotics.sourceforge.net
```
Update  the website.

bvadorno commented on robots/FrankaEmikaPandaRobot.m:
```diff
@@ -0,0 +1,61 @@
+classdef FrankaEmikaPandaRobot
+    methods (Static)
+        function ret = kinematics()
+            %Modified D-H of FrankaEmikaPanda
```
What's the justification for using the modified DH convention?

juanjqo replied:
I used it since the [official documentation](https://frankaemika.github.io/docs/control_parameters.html) provided by the manufacturer Franka Emika uses the modified DH parameters. On the other hand, I modeled the robot using the standard convention to check the class DQ_SerialManipulatorMDH. So, we have both models available.

bvadorno reviewed (changes requested):
I haven't seen the `DQ_SerialManipulatorMDH` yet because there are too many adjustments in the `DQ_SerialManipulatorDH` that need to be replicated in the MDH file. Please address not only the requested changes but also the clarifications I asked for.

juanjqo commented on robot_modeling/DQ_Kinematics.m:
```diff
@@ -1,10 +1,15 @@
-%   reference_frame - Reference frame (not always coincident with base_frame).
+%   reference_frame_ - Reference frame (not always coincident with base_frame).
 %   q - Robot configuration vector.
+%   dim_configuration_space - Dimension of the robot configuration space.
```
Indeed, we don't. It was a typo.

juanjqo commented on robot_modeling/DQ_Kinematics.m:
```diff
@@ -125,6 +140,27 @@ function set_base_frame(obj, base_frame)
     end
     
+    methods (Access = protected)        
+        function check_to_ith_link(obj, to_ith_link)
```
I implemented the method using the same name as in C++.  @mmmarinho do you have any comments?

mmmarinho replied:
@juanjqo @bvadorno 
The local variable it was checking was called ‘to_ith_link’, first used in methods such as ‘fkm(q,to_ith_link)’ and ‘pose_jacobian(q,to_ith_link)’. These follow Bruno’s logic, because, for example, the local variable refers to the FKM up to a given link. 

The naming might not be the same as in MATLAB. 

One way or the other, it might be a good chance to unify those even though they are all private/protected  or local. 

@juanjqo  If in the C++ version the leading _ is missing on the method signature, please also add it there.

juanjqo commented on robot_modeling/DQ_SerialManipulator.m:
```diff
@@ -45,16 +60,17 @@
-        effector;
+        %theta,d,a,alpha;
+        %convention;        
+        curr_effector_;
```
@bvadorno The variable curr_effector_ is supposed to be private. Your assumption about the convention is right. Therefore, the implementation is wrong. I fixed it.

juanjqo commented on robot_modeling/DQ_SerialManipulator.m:
```diff
@@ -45,16 +60,17 @@
-        effector;
+        %theta,d,a,alpha;
+        %convention;        
+        curr_effector_;
```
Since the main goal of this PR was to update the Matlab version with respect to the version of C++, I implemented the variables using the same name as in C++. I do not know the motivation for _curr_effector__. @mmmarinho do you have any comments about it?

mmmarinho replied:
@juanjqo  It should be effector_ in both languages. Style-wise it should be pared with the setter and getter methods. 

curr_effector_ is probably a remnant of an older implementation.

juanjqo replied:
@dqrobotics/developers I modified  curr_effector_ by effector_.

juanjqo commented on robot_modeling/DQ_SerialManipulator.m:
```diff
@@ -152,188 +232,34 @@ function set_effector(obj,effector)
             % values are 1).
             
-            if nargin == 3 && ith < obj.n_links
+            if nargin == 3 && ith < obj.dim_configuration_space_
```
Again, I implemented it as in the C++ version. But, If I am not wrong about @mmmarinho's logic, in that line of the code, we want to know if the index (valid or not) is lower than the robot configuration space. We are checking if both the joint positions and the index are valid in the concrete classes (DQ_SerialManipulatorDH and DQ_SerialManipulatorMDH).
For instance, if we have: 
```Matlab
robot = KukaLwr4Robot.kinematics;
q = [0 0 0 0 0 0 0];
J = robot.pose_jacobian(q, 7);
```
The method _obj.pose_jacobian(obj, q, 7)_ will run the second condition (the code below the statement _else_) despite the method having 3 arguments and using a valid index.

@mmmarinho if I am wrong, please correct me.

mmmarinho replied:
@juanjqo 
That specific line of code is unique to MATLAB... So not much to do with the CPP implementation IMO.

`ith` will not always exist in the context of the method, because it will depend on which signature you call.
`pose_jacobian(q)` will have `nargin<3`
`pose_jacobian(q,7)` for instance will have `nargin==3`

You can do as @bvadorno  suggested, and do something like

```matlab
if nargin == 3
 _check_ith_link(ith)
%do whatever
endif
```

juanjqo commented on robot_modeling/DQ_SerialManipulatorDH.m:
```diff
@@ -50,10 +52,17 @@
 %
 % Contributors to this file:
 %     Murilo M. Marinho - murilo@nml.t.u-tokyo.ac.jp
+%     Juan Jose Quiroz Omana - juanjqo@g.ecc.u-tokyo.ac.jp
```
@bvadorno that was my fault. In the beginning, I assumed that since I was creating a new m-file I would be the only contributor, even if the code of the new file was based on another implementation.  I made the same error when I implemented the class DQ_SerialManipulatorMDH in C++ and @mmmarinho pointed out the same question. After a brief internal discussion, I understood my error. I made the respective corrections in C++. However, the respective correction in Matlab was missing. I updated the list of contributors.
I think that we can add a comment about that in the CONTRIBUTING.md file to avoid any future misunderstandings. what do you think?

juanjqo commented on robot_modeling/DQ_SerialManipulatorDH.m:
```diff
@@ -19,14 +19,16 @@
+%       get_as - Returns the third row of the DH Matrix ('a' parameter).
+%       get_alphas - Returns the fourth row of the DH Matrix ('alpha' parameter).
+%       get_types - Returns the actuation type, either DQ_SerialManipulatorDH.JOINT_ROTATIONALor DQ_SerialManipulatorDH.JOINT_PRISMATIC
 %       pose_jacobian_derivative - Compute the time derivative of the pose Jacobian.
```
Indeed, it is. The new implementation in cpp-developers, which contains the developments in [robot dynamics ](https://doi.org/10.1115/1.4054320) has the pose_jacobian_derivative method in DQ_Kinematics as an abstract method. However, concrete implementations are not available yet for some classes.

juanjqo replied:
Update:
In this accepted [PR](https://github.com/dqrobotics/cpp/pull/48) for C++ (master branch) the method pose_jacobian_derivative was defined in the superclass DQ_Kinematics.

juanjqo commented on robot_modeling/DQ_SerialManipulatorDH.m:
```diff
@@ -19,14 +19,16 @@
+%       get_ds - Returns the second row of the DH Matrix ('d' parameter).
+%       get_as - Returns the third row of the DH Matrix ('a' parameter).
+%       get_alphas - Returns the fourth row of the DH Matrix ('alpha' parameter).
+%       get_types - Returns the actuation type, either DQ_SerialManipulatorDH.JOINT_ROTATIONALor DQ_SerialManipulatorDH.JOINT_PRISMATIC
```
I modified the name from JOINT_ROTATIONAL and JOINT_PRISMATIC to REVOLUTE AND PRISMATIC, respectively.
@mmmarinho do you think that "JOINT" is necessary for the name?

mmmarinho replied:
@juanjqo @bvadorno If MATLAB has named enums, then we can use that instead. If it doesn’t… then let’s think about it a bit. 

In my opinion, just PRISMATIC or just ROTATIONAL lacks meaning. In C++ I would write DQSerialManipulatorDH::JointType::Rotational with an enum class.

If those are not available, having JOINT_ first is a common style in C enums because it didn’t (or still doesn’t, not sure) have typed enums. Then all related enums were grouped together by their prefix, see OpenCV for one quick example
https://docs.opencv.org/3.4/d9/d6a/group__aruco.html.

mmmarinho replied:
@juanjqo @bvadorno 

I prefer the enumeration class and use it profusely everywhere else. So I'm happy with this change.

Either way, because that old-style enum was available in the last release version (20.04)
https://github.com/dqrobotics/matlab/blob/f11f0de06784d04042b8713010373d08fb86ccf8/robot_modeling/DQ_SerialManipulatorDH.m#L62
we should keep it as deprecated for at least one version, as per our current legacy policy. All examples should be updated to use the new style.

juanjqo replied:
@bvadorno @mmmarinho I added again the properties JOINT_ROTATIONAL and JOINT_PRISMATIC.

juanjqo commented on robot_modeling/DQ_SerialManipulatorDH.m:
```diff
@@ -160,13 +116,15 @@
             
             % The optimized standard dh2dq calculation
```
I did not add that comment. I think that the comment was added in the [first commit](https://github.com/dqrobotics/matlab/commit/fafd4dad68804ffd1f0c3ab99ed315bf35adfda6) of the DQ_SerialManipulatorDH.m (in the master branch) by @mmmarinho.  
@mmmarinho, assuming that I am not wrong, would you answer the question about the "optimized calculation"?

mmmarinho replied:
@bvadorno @juanjqo 

These comments were in the last release version (20.04).
https://github.com/dqrobotics/matlab/blob/f11f0de06784d04042b8713010373d08fb86ccf8/robot_modeling/DQ_SerialManipulatorDH.m#L183

All comments might be removed without any loss, I believe. Especially if they are causing confusion.

Anyways, the "optimized calculation" is the [simplified](https://www.mathworks.com/help/symbolic/simplify.html) form of the `dh2dq` transformation that can be used instead of multiplying the four dual quaternions in sequence, e.g. `h1*h2*h3*h4`. The half angles are a small additional detail.

juanjqo commented on robot_modeling/DQ_SerialManipulatorDH.m:
```diff
@@ -160,13 +116,15 @@
             % Add the effect of the joint value
-            if obj.type(ith) == obj.JOINT_ROTATIONAL
+            %if obj.type(ith) == obj.JOINT_ROTATIONAL
+            if joint_type == obj.JOINT_ROTATIONAL   
```
I am a little confused since those lines of code were not added by me and were not proposed in the current PR. But I understand your comments. I'm going to implement the modifications aiming for better readability and avoiding recalculations.

mmmarinho replied:
@bvadorno @juanjqo 

It's difficult for me to know what is the problem in question because all the files under discussion changed so much, but I believe we might be talking about the modifications for 20.04 discussed here:
https://github.com/dqrobotics/matlab/pull/56#issuecomment-619643404

More specifically, after review 
`3.1 Could you also break the lines at 80 characters to comply with our current guide style?`
the number of characters per line was adjusted but it would have been better to adjust the style to the current usage in MATLAB.

bvadorno commented on robot_modeling/DQ_SerialManipulatorDH.m:
```diff
@@ -19,14 +19,16 @@
-%       fkm - Compute the forward kinematics while taking into account base and end-effector's rigid transformations.
-%       plot - Plots the serial manipulator.
-%       pose_jacobian - Compute the pose Jacobian while taking into account base's and end-effector's rigid transformations.
+%       get_thetas - Returns the first row of the DH Matrix ('theta' parameter). 
```
Hello @juanjqo,

I agree because we need to maintain consistency. In any case, because a robot modeled using DH parameters can have either revolute or prismatic joints, $\theta_i$ and $d_i$ can be either a variable or a parameter. Do we have a function that gives us the configuration vector?

juanjqo replied:
Hi @bvadorno, as far as I know, we do not have such a function. I know that in the superclass DQ_Kinematics we have the protected property _q_. But we are not using that property (at least in the classes related to serial manipulators). The configuration vector is used as an argument in some methods.

bvadorno replied:
In that case, proceed with my previous suggestion. If we didn't need an explicit get method to retrieve the robot configuration until now, implementing it is not urgent.

mmmarinho replied:
@juanjqo @bvadorno 

For either revolute or prismatic joints, the $\theta_i$ or $d_i$ are the offsets of the respective DH parameter. The effective displacement will be $\theta_i+q_i$ or $d_i + q_i$ depending on the joint type.
Until now, the configuration vector has never been stored in the kinematics object. It has always been an argument for the method in question.

@juanjqo changed the internal implementation in MATLAB to look like CPP (I guess to make the adjustments easier?), so some changes got a bit difficult for us to track.

@bvadorno Up until the last release version, `a`, `alpha` etc were class properties without access control
https://github.com/dqrobotics/matlab/blob/f11f0de06784d04042b8713010373d08fb86ccf8/robot_modeling/DQ_SerialManipulator.m#L55
so the user could freely get and set the values of the DH parameters, such as `a` with `robot.a(2)`.

That was (probably?) unintended, so we don't necessarily need to give support to that anymore. This is also where the `get_as()` came from. `get` because we introduced getters and plural because it does not return a single number.

If 
`double DQ_SerialManipulatorDH::get_a(int)`
`VectorXd DQ_SerialManipulatorDH::get_as()` 
are not informative enough, we can do something like:
`double DQ_SerialManipulatorDH::get_dh_parameter_a(int) const`
`VectorXd DQ_SerialManipulatorDH::get_dh_parameter_as() const`

The `_vector()` suffix might also be understood as an implicit commitment to (re)naming methods that return vectors in with the same suffix. In other libraries and interfaces, this is usually dealt with by using plural, for instance `singularValues` in Eigen3
https://eigen.tuxfamily.org/dox/classEigen_1_1JacobiSVD.html
Or `get_joint_positions()` in `DQ_VrepInterface` etc.

 Anyways IMO it is important to be clear and not use plural in a method signature if it is not returning a vector.
 `VectorXd get_parameter_a()` which returns all `a`s seems confusing to me.

juanjqo replied:
@bvadorno @mmmarinho I added two new methods trying to follow the @mmmarinho's suggestion. To do so, I added a new property:
```Matlab
    properties (Constant)
        parameter = struct('theta', 1, 'd',2, 'a', 3, 'alpha', 4, 'JointType', 5);
    end
```
The methods look like this:
```Matlab
        function ret = get_parameters(obj, parameter)
            %GET_PARAMETERS(parameter) Returns the vector containing the
            % parameter specified. The parameter can be theta, a, d, alpha, 
            % and JointType
            ret = obj.dh_matrix_(parameter,:);
        end

        function ret = get_parameter(obj, parameter, ith)
            %GET_PARAMETERS(parameter, ith) Returns the scalar containing the
            % parameter specified of the ith joint. 
            % The parameter can be theta, a, d, alpha, and JointType
            ret = obj.dh_matrix_(parameter,ith);
        end
```
Usage:
```Matlab
robot = DQ_SerialManipulatorDH(ax18_DH_matrix); % Defines robot model using dual quaternions

robot.get_parameters(robot.parameter.a)
robot.get_parameter(robot.parameter.a, 3)

robot.get_parameters(robot.parameter.alpha)
robot.get_parameter(robot.parameter.alpha, 2)
```
What do you think about this?

juanjqo replied:
@mmmarinho @bvadorno Using the same strategy to implement the get_parameters method, I added a new struct property called JointType. We have now the following constant properties
```Matlab
    properties (Constant)
        % Joints that can be actuated
        % Revolute joint
        JOINT_ROTATIONAL = 1; % Deprecated
        % Prismatic joint
        JOINT_PRISMATIC = 2;   % Deprecated
   
        Parameter = struct('Theta', 1, 'D',2, 'A', 3, 'Alpha', 4, 'JointType', 5);    % New
        JointType = struct('REVOLUTE',1, 'PRISMATIC',2);                              % New
    end
```
Usage:

```Matlab
ax18_DH_theta   = [0       0      -pi/2   -pi/2      -pi/2];  % theta
            ax18_DH_d       = [0.167   0       0     0.1225          0];  % d
            ax18_DH_a       = [0      0.159    0.02225    0      0.170];  % a
            ax18_DH_alpha   = [-pi/2   0      -pi/2   -pi/2          0];  % alpha
            ax18_DH_type    = [repmat(DQ_SerialManipulatorDH.JointType.REVOLUTE,1,5)];
            
            ax18_DH_matrix = [  ax18_DH_theta;
                                               ax18_DH_d;
                                               ax18_DH_a;
                                               ax18_DH_alpha;
                                               ax18_DH_type;
               ]; 
    
robot = DQ_SerialManipulatorDH(ax18_DH_matrix);
```

bvadorno commented:
Oh, by the way, @juanjqo, may you include the aforementioned information (i.e., items 1, 2, and 3) in the instructions for pull requests? I guess a good place would be the README.md files in each repository.

juanjqo commented:
> Oh, by the way, @juanjqo, may you include the aforementioned information (i.e., items 1, 2, and 3) in the instructions for pull requests? I guess a good place would be the README.md files in each repository.

@bvadorno Sure!

juanjqo commented:
> Hi @juanjqo,
> 
> @mmmarinho and I concluded that this pull request is unfeasible to double-check and has several elements that are likely to be rejected.
> 
> Therefore,
> 
> 1. restart the procedure with individual changes (several changes of the **same type** are allowed on the same pull request);
> 2. do not unnecessarily change any internal implementation that is working correctly without prior approval;
> 3. and include a clear and concise rationale behind each pull request.
> 
> Many thanks, Bruno

@bvadorno I'll do that!
